pkg/topology/builder.go

```diff
@@ -0,0 +1,667 @@
+				Pods:           destPods,
+			},
+			Specs: specs,
+		}
+		svc.TrafficTargets = append(svc.TrafficTargets, svcTT)
+
+		// Add the ServiceTrafficTarget to the source pods.
+		for _, source := range sources {
+			for _, pod := range source.Pods {
+				pod.Outgoing = append(pod.Outgoing, svcTT)
+			}
+		}
+
+		// Add the ServiceTrafficTarget to the destination pods
+		for _, pod := range svcTT.Destination.Pods {
+			pod.Incoming = append(pod.Incoming, svcTT)
+		}
+	}
+
+	return nil
+}
+
+// evaluateTrafficSplit evaluates the given traffic-split. If the traffic-split targets a known Service, a new TrafficSplit
+// will be added to it. The TrafficSplit will be added only if all its backends expose the ports required by the Service.
+func (b *Builder) evaluateTrafficSplit(topology *Topology, trafficSplit *v1alpha2.TrafficSplit) error {
+	svcKey := NameNamespace{trafficSplit.Spec.Service, trafficSplit.Namespace}
+
+	svc, ok := topology.Services[svcKey]
+	if !ok {
+		return fmt.Errorf("unable to find root Service %s/%s", trafficSplit.Namespace, trafficSplit.Spec.Service)
+	}
+
+	ts := &TrafficSplit{
+		Name:      trafficSplit.Name,
+		Namespace: trafficSplit.Namespace,
+		Service:   svc,
+	}
+
+	backends := make([]TrafficSplitBackend, len(trafficSplit.Spec.Backends))
+
+	for i, backend := range trafficSplit.Spec.Backends {
+		backendSvcKey := NameNamespace{backend.Service, trafficSplit.Namespace}
+
+		backendSvc, ok := topology.Services[backendSvcKey]
+		if !ok {
+			return fmt.Errorf("unable to find backend Service %s/%s", trafficSplit.Namespace, backend.Service)
+		}
+
+		// As required by the SMI specification, backends must expose at least the same ports as the Service on
+		// which the TrafficSplit is.
+		for _, svcPort := range svc.Ports {
+			var portFound bool
+
+			for _, backendPort := range backendSvc.Ports {
+				if svcPort.Port == backendPort.Port {
+					portFound = true
+					break
+				}
+			}
+
+			if !portFound {
+				return fmt.Errorf("port %d must be exposed by Service %s/%s in order to be used as a backend", svcPort.Port, backendSvc.Namespace, backendSvc.Name)
+			}
+		}
+
+		backends[i] = TrafficSplitBackend{
+			Weight:  backend.Weight,
+			Service: backendSvc,
+		}
+
+		backendSvc.BackendOf = append(backendSvc.BackendOf, ts)
+	}
+
+	ts.Backends = backends
+	svc.TrafficSplits = append(svc.TrafficSplits, ts)
+
+	return nil
+}
+
+// populateTrafficSplitsAuthorizedIncomingTraffic computes the list of pods allowed to access a traffic-split. As
+// traffic-splits may form a graph, it has to be done once all the traffic-splits have been processed. To avoid runtime
+// issues, this method detects cycles in the graph.
+func (b *Builder) populateTrafficSplitsAuthorizedIncomingTraffic(topology *Topology) {
+	loopDetected := make(map[*Service][]*TrafficSplit)
+
+	for _, svc := range topology.Services {
+		for _, ts := range svc.TrafficSplits {
+			pods, err := b.getIncomingPodsForTrafficSplit(ts, map[NameNamespace]bool{})
```

Comment:
Why are you passing a new empty map to the function instead of creating it at the beginning of the func? 🤔 If you were reusing it between function calls I'd understand but this seems like a mistake.

Also, when storing such information as whether a name+namespace combo has been visited already or not, since you will never store `false`, using a `bool` here is overkill. Using a `struct{}` is actually more explicit here.

([Brought to you by Dave](https://dave.cheney.net/2014/03/25/the-empty-struct))

Reply:
I agree for the `struct{}` :+1: 
The `map[NameNamespace]struct{}` represents the visited nodes when traversing the graph to detect cycles. It must be a new map each time we call this function because we are interested in detecting cycles on the given TrafficSplit.

Reply:
It can be a new map each time by creating it at the beginning of `getIncomingPodsForTrafficSplit` instead of always passing an empty map as a parameter, is what I meant! Unless there is a case where `getIncomingPodsForTrafficSplit` does not take an empty map?

Reply:
``getIncomingPodsForTrafficSplit` is called by `populateTrafficSplitsAuthorizedIncomingTraffic` and the `visited` map has to be a new one then, `getIncomingPodsForTrafficSplit` is also called recursively in `getIncomingPodsForService`. This second must not create a new `visited` map as we are interested in previously visited nodes to detect cycles.